Nice:

```

gnrc_netif.c: In function ‘_netif_set_addr’:
gnrc_netif.c:1315:23: error: ‘l2util_addr_from_str’ accessing 8 bytes in a region of size 6 [-Werror=stringop-overflow=]
 1315 |     size_t addr_len = l2util_addr_from_str(addr_str, addr);
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
gnrc_netif.c:1315:23: note: referencing argument 2 of type ‘uint8_t *’ {aka ‘unsigned char *’}
In file included from ../../include/net/gnrc/netif.h:62,
                 from ../../include/net/gnrc.h:294,
                 from gnrc_netif.c:28:
../../include/net/l2util.h:192:8: note: in a call to function ‘l2util_addr_from_str’
  192 | size_t l2util_addr_from_str(const char *str, uint8_t out[L2UTIL_ADDR_MAX_LEN]);
      |        ^~~~~~~~~~~~~~~~~~~~
```

So the check for calls with too small target buffer does work.

--- a/sys/include/net/l2util.h
+++ b/sys/include/net/l2util.h
@@ -27,6 +27,7 @@
 
 #include <stdint.h>
 
+#include "compiler_hints.h"
 #include "net/eui64.h"
 #include "net/ndp.h"
 
@@ -177,18 +178,41 @@ char *l2util_addr_to_str(const uint8_t *addr, size_t addr_len, char *out);
  * @details The input format must be like `xx:xx:xx:xx` where `xx` will be the
  *          bytes of @p addr in hexadecimal representation.
  *
- * @pre `(out != NULL)`
- * @pre @p out **MUST** have allocated at least
- *      @ref GNRC_NETIF_L2ADDR_MAXLEN bytes.
+ * @pre     `(out != NULL)`
+ * @pre     @p out **MUST** have allocated at least @ref L2UTIL_ADDR_MAX_LEN
+ *          bytes.
+ * @pre     @p str must be `\0` terminated
  *
- * @param[in] str       A string of colon-separated hexadecimals.
- * @param[out] out      The resulting hardware address. Must at least have
+ * @param[in]   str         A string of colon-separated hexadecimals.
+ * @param[out]  addr        The resulting hardware address.
+ * @param[in]   addr_size   Size of @p addr
+ *
+ * @return  Actual length of @p out on success (may be less than @p addr_size)
+ * @retval  0           Invalid input
+ */
+ACCESS(write_only, 2, 3)
+size_t l2util_addr_from_str_sized(const char *str, void *addr, size_t addr_size);
+
+/**
+ * @brief   Parses a string of colon-separated hexadecimals to a hardware
+ *          address.
+ *
+ * @details The input format must be like `xx:xx:xx:xx` where `xx` will be the
+ *          bytes of @p addr in hexadecimal representation.
+ *
+ * @pre     `(out != NULL)`
+ * @pre     @p out **MUST** have allocated at least @ref L2UTIL_ADDR_MAX_LEN
+ *          bytes.
+ * @pre     @p str must be `\0` terminated
+ *
+ * @param[in]   str     A string of colon-separated hexadecimals.
+ * @param[out]  out     The resulting hardware address. Must at least have
  *                      @ref GNRC_NETIF_L2ADDR_MAXLEN bytes allocated.
  *
  * @return  Actual length of @p out on success.
- * @return  0, on failure.
+ * @retval  0           Invalid input
  */
-size_t l2util_addr_from_str(const char *str, uint8_t *out);
+size_t l2util_addr_from_str(const char *str, uint8_t out[L2UTIL_ADDR_MAX_LEN]);
 
 /**
  * @brief   Checks if two l2 addresses are equal.

--- a/sys/net/link_layer/l2util/l2util.c
+++ b/sys/net/link_layer/l2util/l2util.c
@@ -350,7 +350,7 @@ char *l2util_addr_to_str(const uint8_t *addr, size_t addr_len, char *out)
     return res;
 }
 
-size_t l2util_addr_from_str(const char *str, uint8_t *out)
+size_t l2util_addr_from_str_sized(const char *str, void *_addr, size_t addr_size)
 {
     /* Walk over str from the end. */
     /* Take two chars a time as one hex value (%hhx). */
@@ -358,46 +358,61 @@ size_t l2util_addr_from_str(const char *str, uint8_t *out)
     /* Every non-hexadimal character is a delimiter. */
     /* Leading, tailing and adjacent delimiters are forbidden. */
     const char *end_str = str;
-    uint8_t *out_end = out;
+    uint8_t *addr = _addr;
+    uint8_t *addr_end = addr;
     size_t count = 0;
     int assert_cell = 1;
 
-    assert(out != NULL);
+    assert(addr != NULL);
     if ((str == NULL) || (str[0] == '\0')) {
         return 0;
     }
+
     /* find end of string */
     while (end_str[1]) {
         ++end_str;
     }
+
     while (end_str >= str) {
-        int a = 0, b = _dehex(*end_str--, -1);
+        int a = 0;
+        int b = _dehex(*end_str--, -1);
 
         if (b < 0) {
             if (assert_cell) {
                 return 0;
             }
-            else {
-                assert_cell = 1;
-                continue;
-            }
+
+            assert_cell = 1;
+            continue;
         }
         assert_cell = 0;
         if (end_str >= str) {
             a = _dehex(*end_str--, 0);
         }
         count++;
-        *out_end++ = (a << 4) | b;
+
+        if (count > addr_size) {
+            return 0;
+        }
+
+        *addr_end++ = (a << 4) | b;
     }
+
     if (assert_cell) {
         return 0;
     }
+
     /* out is reversed */
-    while (out < --out_end) {
-        uint8_t tmp = *out_end;
-        *out_end = *out;
-        *out++ = tmp;
+    while (addr < --addr_end) {
+        uint8_t tmp = *addr_end;
+        *addr_end = *addr;
+        *addr++ = tmp;
     }
     return count;
+}
+
+size_t l2util_addr_from_str(const char *str, uint8_t out[L2UTIL_ADDR_MAX_LEN])
+{
+    return l2util_addr_from_str_sized(str, out, L2UTIL_ADDR_MAX_LEN);
 }
 /** @} */

--- a/sys/test_utils/netdev_ieee802154_minimal/shell_commands.c
+++ b/sys/test_utils/netdev_ieee802154_minimal/shell_commands.c
@@ -189,7 +189,7 @@ static int cmd_txtsnd(int argc, char **argv)
     case 4:
         break;
     case 5:
-        res = l2util_addr_from_str(argv[idx++], pan.u8);
+        res = l2util_addr_from_str_sized(argv[idx++], pan.u8, sizeof(pan));
         if ((res == 0) || (res > sizeof(pan))) {
             txtsnd_usage(argv[0]);
             return 1;

--- a/tests/net/l2util/main.c
+++ b/tests/net/l2util/main.c
@@ -446,7 +446,9 @@ static void test_l2util_addr_from_str(void)
     static const uint8_t ethernet_l2addr[] = ETHERNET_SRC;
     static const uint8_t ieee802154_l2addr_long[] = IEEE802154_LONG_SRC;
     static const uint8_t ieee802154_l2addr_short[] = IEEE802154_SHORT_SRC;
-    uint8_t out[GNRC_NETIF_L2ADDR_MAXLEN];
+    uint8_t out[L2UTIL_ADDR_MAX_LEN + 1];
+    const uint8_t canary = 0x67;
+    out[L2UTIL_ADDR_MAX_LEN] = canary;
 
     TEST_ASSERT_EQUAL_INT(0, l2util_addr_from_str("", out));
     TEST_ASSERT_EQUAL_INT(sizeof(ethernet_l2addr),
@@ -458,10 +460,17 @@ static void test_l2util_addr_from_str(void)
                                                    out));
     TEST_ASSERT_EQUAL_INT(0, memcmp(ieee802154_l2addr_long, out,
                                     sizeof(ieee802154_l2addr_long)));
+    TEST_ASSERT_EQUAL_INT(0,
+                          l2util_addr_from_str_sized("3E:E6:B5:0F:19:22:FD:0A",
+                                                     out,
+                                                     6));
     TEST_ASSERT_EQUAL_INT(sizeof(ieee802154_l2addr_short),
                           l2util_addr_from_str("FD:0A", out));
     TEST_ASSERT_EQUAL_INT(0, memcmp(ieee802154_l2addr_short, out,
                                     sizeof(ieee802154_l2addr_short)));
+    TEST_ASSERT_EQUAL_INT(0, l2util_addr_from_str("01:02:03:04:05:06:07:08:09",
+                                                  out));
+    TEST_ASSERT_EQUAL_INT(canary, out[L2UTIL_ADDR_MAX_LEN]);
 }
 
 TestRef test_l2util(void)

